File: ir/function_ir.py
from common.span import SourceSpan
from .annotation_ir import AnnotationIR
from .ir_node import IRNode
from generated import _pb2
from .expr_ir import ExprIR
from .stmt_ir import StmtIR
from .decl_ir import DeclIR


class ParamIR(IRNode):
    def __init__(
        self,
        symbol_id: int,
        name: str,
        annotation: AnnotationIR,
        default,
        span: SourceSpan,
    ):
        self.symbol_id = symbol_id
        self.name = name
        self.annotation = annotation
        self.default = default
        self.span = span

    def to_proto(self):
        proto = _pb2.ParamIR(
            symbol_id=self.symbol_id,
            name=self.name,
        )

        if self.annotation is not None:
            proto.annotation.CopyFrom(self.annotation.to_proto())

        if self.default is not None:
            proto.default_value.CopyFrom(self.default.to_proto())

        if self.span is not None:
            proto.span.CopyFrom(self.span.to_proto())

        return proto

class ReturnIR(StmtIR):
    def __init__(self, value: ExprIR, span: SourceSpan):
        super().__init__(span=span)
        self.span = span
        self.value = value

    def to_proto(self):
        proto = _pb2.ReturnIR()

        if self.value is not None:
            proto.value.CopyFrom(self.value.to_proto())

        if self.span is not None:
            proto.span.CopyFrom(self.span.to_proto())

        stmt = _pb2.StmtIR()
        stmt.return_stmt.CopyFrom(proto)
        return stmt


class FunctionIR(DeclIR):
    def __init__(
        self,
        id: int,  # symbol id -> name of function
        symbol_id: int,
        name: str,
        scope_id: int,  # parent scope where function name is bound
        body_scope_id: int,  # function-local scope
        params: list[ParamIR],
        body: list[IRNode],
        returns,
        decorators,
        span: SourceSpan,
    ):
        super().__init__(span=span)
        self.id = id
        self.symbol_id = symbol_id
        self.name = name
        self.scope_id = scope_id
        self.body_scope_id = body_scope_id
        self.params = params
        self.body = body
        self.returns = returns
        self.decorators = decorators
        self.span = span

    def to_proto(self):
        proto = _pb2.FunctionIR(
            id=self.id,
            symbol_id=self.symbol_id,
            name=self.name,
            scope_id=self.scope_id,
            body_scope_id=self.body_scope_id,
        )

        proto.params.extend([p.to_proto() for p in self.params])
        proto.body.extend([stmt.to_proto() for stmt in self.body])
        proto.decorators.extend([d.to_proto() for d in self.decorators])

        if self.returns is not None:
            proto.returns.CopyFrom(self.returns.to_proto())

        if self.span is not None:
            proto.span.CopyFrom(self.span.to_proto())

        stmt = _pb2.DeclIR()
        stmt.function.CopyFrom(proto)
        return stmt

File: rust/src/ir/stmt/function_ir.rs
use crate::ir::nodes::annotation_ir::AnnotationIR;
use crate::ir::{expr_ir::ExprIR, span_ir::SourceSpan, stmt_ir::StmtIR};

#[derive(Debug, Clone)]
pub enum ParamKind {
    PositionalOnly = 1,
    PositionalOrKeyword = 2,
    VarPositional = 3,
    KeywordOnly = 4,
    VarKeyword = 5,
}

impl TryFrom<i32> for ParamKind {
    type Error = String;

    fn try_from(int: i32) -> Result<Self, Self::Error> {
        match int {
            1 => Ok(ParamKind::PositionalOnly),
            2 => Ok(ParamKind::PositionalOrKeyword),
            3 => Ok(ParamKind::VarPositional),
            4 => Ok(ParamKind::KeywordOnly),
            5 => Ok(ParamKind::VarKeyword),
            _ => Err(format!("invalid ParamKind value: {}", int)),
        }
    }
}

#[derive(Debug, Clone)]
pub struct ParamIR {
    pub symbol_id: i64,
    pub name: String,
    pub kind: ParamKind,
    pub annotation: Option<AnnotationIR>,
    pub default: Option<Box<ExprIR>>,
    pub span: Option<SourceSpan>,
}

impl ParamIR {
    pub fn new(
        symbol_id: i64,
        name: impl Into<String>,
        kind: ParamKind,
        annotation: Option<AnnotationIR>,
        default: Option<ExprIR>,
        span: Option<SourceSpan>,
    ) -> Self {
        Self {
            symbol_id,
            name: name.into(),
            kind,
            annotation,
            default: default.map(Box::new),
            span,
        }
    }
}

#[derive(Debug, Clone)]
pub struct ReturnIR {
    pub value: Option<Box<ExprIR>>,
    pub span: Option<SourceSpan>,
}

impl ReturnIR {
    pub fn new(value: Option<ExprIR>, span: Option<SourceSpan>) -> Self {
        Self {
            value: value.map(Box::new),
            span,
        }
    }
}

#[derive(Debug, Clone)]
pub struct FunctionIR {
    pub id: i64,
    pub symbol_id: i64,
    pub name: String,
    pub scope_id: i64,
    pub body_scope_id: i64,
    pub params: Vec<ParamIR>,
    pub body: Vec<StmtIR>,
    pub returns: Option<AnnotationIR>,
    pub decorators: Vec<ExprIR>,
    pub span: Option<SourceSpan>,
}
